libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,794 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "tenzir/arrow_time_utils.hpp"
+#include "tenzir/arrow_utils.hpp"
+#include "tenzir/cast.hpp"
+#include "tenzir/concept/printable/tenzir/json.hpp"
+#include "tenzir/detail/assert.hpp"
+#include "tenzir/detail/enumerate.hpp"
+#include "tenzir/detail/zip_iterator.hpp"
+#include "tenzir/series_builder.hpp"
+#include "tenzir/tql2/ast.hpp"
+#include "tenzir/tql2/eval.hpp"
+#include "tenzir/tql2/plugin.hpp"
+#include "tenzir/tql2/resolve.hpp"
+
+namespace tenzir::plugins::chart {
+
+TENZIR_ENUM(chart_type, area, bar, line, pie);
+
+namespace {
+
+using bucket = std::vector<std::unique_ptr<aggregation_instance>>;
+// Will point to valid strings, as `std::unordered_set` does not invalidate
+// pointers on insertion
+using grouped_bucket = detail::stable_map<std::string_view, bucket>;
+using group_map = std::map<data, grouped_bucket>;
+using call_map = detail::stable_map<std::string, ast::function_call>;
+using plugins_map
+  = std::unordered_multimap<const aggregation_plugin*, ast::function_call>;
+
+struct xlimit {
+  located<data> value;
+  expression legacy_expr;
+  ast::expression expr;
+
+  friend auto inspect(auto& f, xlimit& x) -> bool {
+    return f.object(x).fields(f.field("value", x.value),
+                              f.field("legacy_expr", x.legacy_expr),
+                              f.field("expr", x.expr));
+  }
+};
+
+struct chart_args {
+  chart_type ty;
+  ast::simple_selector x;
+  call_map y;
+  std::optional<ast::expression> group;
+  std::optional<xlimit> xmin;
+  std::optional<xlimit> xmax;
+  std::optional<located<data>> ymin;
```

Comment:
Can you check that the type of `y_min` and `y_max` matches the output type? Otherwise weird things happen:
<img width="1160" alt="image" src="https://github.com/user-attachments/assets/bfa40397-c3bb-4594-a2f6-aedd99a9d365" />

Reply:
Added check.